flink-runtime/src/main/java/org/apache/flink/runtime/accumulators/AccumulatorRegistry.java

```diff
@@ -0,0 +1,144 @@
+import org.apache.flink.runtime.executiongraph.ExecutionAttemptID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+
+/**
+ * Main accumulator registry which encapsulates internal and user-defined accumulators.
+ */
+public class AccumulatorRegistry {
+
+	protected static final Logger LOG = LoggerFactory.getLogger(AccumulatorRegistry.class);
+
+	protected final JobID jobID;
+	protected final ExecutionAttemptID taskID;
+
+	/* Flink's internal Accumulator values stored for the executing task. */
+	private final Map<Metric, Accumulator<?, ?>> flinkAccumulators =
+			new HashMap<Metric, Accumulator<?, ?>>();
+
+	/* User-defined Accumulator values stored for the executing task. */
+	private final Map<String, Accumulator<?, ?>> userAccumulators =
+			Collections.synchronizedMap(new HashMap<String, Accumulator<?, ?>>());
+
+	/* The reporter returned to reporting tasks. */
+	private final ReadWriteReporter reporter;
+
+	/**
+	 * Flink metrics supported
+	 */
+	public enum Metric {
+		NUM_RECORDS_IN,
+		NUM_RECORDS_OUT,
+		NUM_BYTES_IN,
+		NUM_BYTES_OUT
+	}
+
+
+	public AccumulatorRegistry(JobID jobID, ExecutionAttemptID taskID) {
+		this.jobID = jobID;
+		this.taskID = taskID;
+		this.reporter = new ReadWriteReporter(flinkAccumulators);
+	}
+
+	/**
+	 * Creates a snapshot of this accumulator registry.
+	 * @return a serialized accumulator map
+	 */
+	public AccumulatorSnapshot getSnapshot() {
+		try {
+			return new AccumulatorSnapshot(jobID, taskID, flinkAccumulators, userAccumulators);
+		} catch (IOException e) {
+			LOG.warn("Failed to serialize accumulators for task.", e);
+			return null;
+		}
+	}
+
+	/**
+	 * Gets the map for user-defined accumulators.
+	 */
+	public Map<String, Accumulator<?, ?>> getUserMap() {
+		return userAccumulators;
+	}
+
+	/**
+	 * Gets the reporter for flink internal metrics.
+	 */
+	public Reporter getReadWriteReporter() {
+		return reporter;
+	}
+
+	/**
+	 * Interface for Flink's internal accumulators.
+	 */
+	public interface Reporter {
+		void report(Metric metric, long value);
+	}
+
+	/**
+	 * Accumulator based reporter for keeping track of internal metrics (e.g. bytes and records in/out)
+	 */
+	public static class ReadWriteReporter implements Reporter {
+
+		private LongCounter numRecordsIn = new LongCounter();
+		private LongCounter numRecordsOut = new LongCounter();
+		private LongCounter numBytesIn = new LongCounter();
+		private LongCounter numBytesOut = new LongCounter();
+
+		public ReadWriteReporter(Map<Metric, Accumulator<?,?>> accumulatorMap) {
+			accumulatorMap.put(Metric.NUM_RECORDS_IN, numRecordsIn);
+			accumulatorMap.put(Metric.NUM_RECORDS_OUT, numRecordsOut);
+			accumulatorMap.put(Metric.NUM_BYTES_IN, numBytesIn);
+			accumulatorMap.put(Metric.NUM_BYTES_OUT, numBytesOut);
+		}
+
+		@Override
+		public void report(Metric metric, long value) {
```

Comment:
This is called extremely often - once per tuple. It needs to have virtually zero overhead.

To achieve that, I would add dedicated methods for each metric, rather than have the enumeration.